client/my-sites/domains/domain-management/components/primary-domain-selector/index.tsx

```diff
@@ -0,0 +1,184 @@
+	}
+
+	const isOnFreePlan = site?.plan?.is_free ?? false;
+
+	const shouldUpgradeToMakeDomainPrimary = ( domain: DomainData ): boolean => {
+		return (
+			! userCanSetPrimaryDomains &&
+			isOnFreePlan &&
+			( domain.type === 'registered' || domain.type === 'mapping' ) &&
+			! domain.current_user_can_create_site_from_domain_only &&
+			! domain.wpcom_domain &&
+			! domain.is_wpcom_staging_domain &&
+			( site?.plan?.features.active.includes( FEATURE_SET_PRIMARY_CUSTOM_DOMAIN ) ?? false )
+		);
+	};
+
+	let validPrimaryDomains = domains.filter( ( domain ) => {
+		return (
+			domain.can_set_as_primary &&
+			! domain.aftermarket_auction &&
+			domain.points_to_wpcom &&
+			! shouldUpgradeToMakeDomainPrimary( domain ) &&
+			! domain.primary_domain
+		);
+	} );
+
+	const hasWpcomStagingDomain = validPrimaryDomains.find(
+		( domain ) => domain.is_wpcom_staging_domain
+	);
+
+	if ( hasWpcomStagingDomain ) {
+		validPrimaryDomains = validPrimaryDomains.filter( ( domain ) => {
+			if ( domain.wpcom_domain ) {
+				return domain.is_wpcom_staging_domain;
+			}
+
+			return true;
+		} );
+	}
+
+	const onSelectChange = ( event: ChangeEvent< HTMLSelectElement > ) => {
+		setSelectedDomain( event.target.value );
+	};
+
+	const onSubmit = ( event: FormEvent< HTMLFormElement > ) => {
+		event.preventDefault();
+
+		const domain = domains.find( ( d ) => d.domain === selectedDomain );
+		if ( ! domain || ! selectedDomain ) {
+			return;
+		}
+
+		setIsSettingPrimaryDomain( true );
+		onSetPrimaryDomain( selectedDomain, () => setIsSettingPrimaryDomain( false ), domain.type );
+	};
+
+	const trackUpgradeClick = () => {
+		recordTracksEvent( 'calypso_primary_site_address_nudge_cta_click', {
+			cta_name: 'add_custom_domain',
+		} );
+	};
+
+	return (
+		<FormFieldset className="domains-set-primary-address">
+			<FormLabel htmlFor="primary-domain-selector" className="domains-set-primary-address__title">
+				{ translate( 'Primary site address' ) }
+			</FormLabel>
+			<div>
+				<FormSettingExplanation className="domains-set-primary-address__subtitle">
+					{ translate(
+						'The current primary address set for this site is: {{strong}}%(selectedDomain)s{{/strong}}.',
+						{
+							args: { selectedDomain: primaryDomain as string },
+							components: {
+								strong: <strong />,
+							},
+						}
+					) }{ ' ' }
+					{ validPrimaryDomains.length > 1
+						? translate(
+								'You can change it by selecting a different address from the list below. {{learnMoreLink}}Learn more{{/learnMoreLink}}.',
+								{
+									components: {
+										learnMoreLink: (
+											<InlineSupportLink supportContext="primary-site-address" showIcon={ false } />
+										),
+									},
+								}
+						  )
+						: translate(
+								'Before you can change the primary site address you have to add a new custom domain. Buy a {{domainSearchLink}}custom domain{{/domainSearchLink}} or {{mapDomainLink}}map{{/mapDomainLink}} a domain you already own.',
```

Comment:
I think that Calypso generally now prefers the term "connect" over "map" - might be wrong on this though?

<img width="1035" alt="Screenshot 2024-03-15 at 18 09 59" src="https://github.com/Automattic/wp-calypso/assets/43215253/66b4d4d0-604b-4047-8e8a-7ce9def1a5f4">

```suggestion
								'Before changing the primary site address, you have to add a new custom domain. Buy a {{domainSearchLink}}custom domain{{/domainSearchLink}} or {{mapDomainLink}}connect{{/mapDomainLink}} a domain you already own.',
```